lib/src/react_client/private_utils.dart

```diff
@@ -0,0 +1,231 @@
+      originalEventHandlers[reactDartConvertedEventHandler] = value;
+    }
+  });
+}
+
+void convertRefValue(Map args) {
+  var ref = args['ref'];
+  if (ref is Ref) {
+    args['ref'] = ref.jsRef;
+  }
+}
+
+void convertRefValue2(Map args, {bool convertCallbackRefValue = true}) {
+  var ref = args['ref'];
+
+  if (ref is Ref) {
+    args['ref'] = ref.jsRef;
+    // If the ref is a callback, pass ReactJS a function that will call it
+    // with the Dart Component instance, not the ReactComponent instance.
+    //
+    // Use CallbackRef<Null> to check arity, since parameters could be non-dynamic, and thus
+    // would fail the `is CallbackRef<dynamic>` check.
+    // See https://github.com/dart-lang/sdk/issues/34593 for more information on arity checks.
+  } else if (ref is CallbackRef<Null> && convertCallbackRefValue) {
+    args['ref'] = allowInterop((dynamic instance) {
+      // Call as dynamic to perform dynamic dispatch, since we can't cast to CallbackRef<dynamic>,
+      // and since calling with non-null values will fail at runtime due to the CallbackRef<Null> typing.
+      if (instance is ReactComponent && instance.dartComponent != null) return (ref as dynamic)(instance.dartComponent);
+      return (ref as dynamic)(instance);
+    });
+  }
+}
+
+/// Util used with `registerComponent2` to ensure no important lifecycle
+/// events are skipped. This includes `shouldComponentUpdate`,
+/// `componentDidUpdate`, and `render` because they utilize
+/// `_updatePropsAndStateWithJs`.
+///
+/// Returns the list of lifecycle events to skip, having removed the
+/// important ones. If an important lifecycle event was set for skipping, a
+/// warning is issued.
+List<String> filterSkipMethods(List<String> methods) {
+  List<String> finalList = List.from(methods);
+  bool shouldWarn = false;
+
+  if (finalList.contains('shouldComponentUpdate')) {
+    finalList.remove('shouldComponentUpdate');
+    shouldWarn = true;
+  }
+
+  if (finalList.contains('componentDidUpdate')) {
+    finalList.remove('componentDidUpdate');
+    shouldWarn = true;
+  }
+
+  if (finalList.contains('render')) {
+    finalList.remove('render');
+    shouldWarn = true;
+  }
+
+  if (shouldWarn) {
+    window.console.warn("WARNING: Crucial lifecycle methods passed into "
+        "skipMethods. shouldComponentUpdate, componentDidUpdate, and render "
+        "cannot be skipped and will still be added to the new component. Please "
+        "remove them from skipMethods.");
+  }
+
+  return finalList;
+}
+
+/// Converts a list of variadic children arguments to children that should be passed to ReactJS.
+///
+/// Returns:
+///
+/// - `null` if there are no args and [shouldAlwaysBeList] is false
+/// - `[]` if there are no args and [shouldAlwaysBeList] is true
+/// - the single child if only one was specified
+/// - otherwise, the same list of args, will all top-level children validated
+dynamic generateChildren(List childrenArgs, {bool shouldAlwaysBeList = false}) {
```

Comment:
Same as above; this seems like it could be private in `component_factory.dart`. Or, maybe since most of the methods in here are factory related, they could all be split out go into a `factory_util.dart`?

Reply:
Ohh I like the factory_util.dart idea!